compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -510,18 +593,53 @@ abstract class GenJSCode extends plugins.PluginComponent
      *  The fields are initialized with the zero of their types.
      */
     def genClassFields(cd: ClassDef): List[js.FieldDef] = {
+      val classSym = cd.symbol
+      assert(currentClassSym.get == classSym,
+          "genClassFields called with a ClassDef other than the current one")
+
       // Non-method term members are fields, except for module members.
       (for {
-        f <- currentClassSym.info.decls
+        f <- classSym.info.decls
         if !f.isMethod && f.isTerm && !f.isModule
       } yield {
         implicit val pos = f.pos
+
         val mutable =
           suspectFieldMutable(f) || unexpectedMutatedFields.contains(f)
-        js.FieldDef(encodeFieldSym(f), toIRType(f.tpe), mutable)
+        val name =
+          if (isExposed(f)) js.StringLiteral(jsNameOf(f))
+          else encodeFieldSym(f)
+
+        val irTpe = {
+          if (f.tpe.typeSymbol != CharClass || !isScalaJSDefinedJSClass(classSym))
+            toIRType(f.tpe)
+          else
+            jstpe.ClassType(ir.Definitions.BoxedCharacterClass)
+        }
```

Comment:
> What is wrong with erasing the field to Any in jsinterop and let scalac handle the rest?

The same as if we erased it right here: an *uninitialized* `var` (with `= _`) will contain the zero of a primitive type (e.g., `null`), but accesses to the field will read it as `obj.field.asInstanceOf[ValueClass].underlyingField` which will NPE at best.

This is because scalac (erroneously IMO) doesn't treat `null` as the 0 of the underlying type when "unboxing" a ValueClass.

Since the field is directly accessible by JS, we cannot control reads of the field to fix up `null` at the time of reading. And hence a `null` can leak out, be passed around, and eventually arrive at an unboxing operation which will NPE (and it shouldn't have, since it should have be an actually boxed zero instead of `null`).